### ISSUE TITLE
Cscope invocation is fragile (and dangerous)

### ISSUE THREAD

**Issue by trou** [vim/vim#3184]
In function `cs_create_connection` from `if_cscope.c`, the following code builds the `cscope` command line and invokes it:

```C
	/* run the cscope command; is there execl for non-unix systems? */
#if defined(UNIX)
	(void)sprintf(cmd, "exec %s -dl -f %s", prog, csinfo[i].fname);
#else
	/* WIN32 */
	(void)sprintf(cmd, "%s -dl -f %s", prog, csinfo[i].fname);
#endif
	if (csinfo[i].ppath != NULL)
	{
	    (void)strcat(cmd, " -P");
	    (void)strcat(cmd, csinfo[i].ppath);
	}
	if (csinfo[i].flags != NULL)
	{
	    (void)strcat(cmd, " ");
	    (void)strcat(cmd, csinfo[i].flags);
	}
# ifdef UNIX
      /* on Win32 we still need prog */
	vim_free(prog);
# endif
	vim_free(ppath);

#if defined(UNIX)
# if defined(HAVE_SETSID) || defined(HAVE_SETPGID)
	/* Change our process group to avoid cscope receiving SIGWINCH. */
#  if defined(HAVE_SETSID)
	(void)setsid();
#  else
	if (setpgid(0, 0) == -1)
	    PERROR(_("cs_create_connection setpgid failed"));
#  endif
# endif
	if (execl("/bin/sh", "sh", "-c", cmd, (char *)NULL) == -1)
	    PERROR(_("cs_create_connection exec failed"));
```

So having a directory with a '(' in the path makes vim cscope invocation fail:
```shell
mkdir '(test)'
cd '(test)'
touch cscope.out
vim --cmd "cs add cscope.out"
```
produces
```
Error detected while processing pre-vimrc command line:
cs_read_prompt EOF: Success
E609: Cscope error: sh: 1: Syntax error: "(" unexpected
```
which is confirmed by strace:
```
[pid 21065] execve("/bin/sh", ["sh", "-c", "exec cscope -dl -f /tmp/(test)/cscope.out"],  /* 45 vars */) = 0
```

This can be abused to execute arbitrary commands (although the exploitation is unlikely as the user will probably see the weird path :)

For example: running vim with <https://github.com/simplyzhao/cscope_maps.vim> in a directory name `` /tmp/`touch pwned ;exit 0`/ `` containing a `cscope.out` file will execute `touch pwned; exit 0`

**Comment by kashperanto** [comment]
I have run into this bug in the wild, and the error message is cryptic enough that it took me some time to find the real issue.

Plenty of people out there do not use CLI tools and are not afraid to use parenthesis in their filenames.

**Comment by chrisbra** [comment]
fixed by 4050305bfd24e7d0e241f617519bd0fa86263b9a
